2.0.x - Switch to U8G2 library - not ready for prime time

Bob-the-Kuhn opened the pull request:
This is a snapshot of the my efforts to switch Marlin 2.0 from the U8G library to the U8G2 library.

The purpose of this PR is to get feedback and hopefully some testing.

The current status is the SPI part is 90% complete.  It displays correctly on LP1768 and on the mega2560.  So far I've tested on the VIKI2, the RepRap Discount Full Graphic controller and on the MKS MINI12864.

Attached are memory and performance testing of 
- U8G2 Re-ARM & mega2560
- Baseline U8G Re-ARM & mega2560
- Marlin 1.15 U8G mega 2560
[U8G2 vs U8G rev 2.zip](https://github.com/MarlinFirmware/Marlin/files/1404576/U8G2.vs.U8G.rev.2.zip)

----

The major changes are:
1. **ultralcd_impl_DOGM.h** - almost all the code changes are in this file.  Mostly it was just changing names of functions.  I did have to add some `setDrawColor` statements to get things to display correctly.  There's a lot of commented out items I need to remove & I need to convert all the WHITE & BLACK labels back to "1" and "0".
2. Add low level SPI drivers for LPC1768
3. Converted fonts to U8G2 format
4. Converted bitmaps to U8G2 format (just required transposing bits within each byte - 0 & 7 swapped, 1 & 6 swapped, ...)
5. Add logical mapping between existing U8G2 drivers and new low level LPC1768 drivers

The new low level LPC1768 drivers may apply to more than just the LPC1768.  There's nothing in there that is LPC1768 specific.

----

I was able to find a way to specify, in the Configuration.h file, the bit rate for the Re-ARM SPI based LCDs.

That method should also allow us to specify the I2C address and speed in the Configuration.h file.

----

Many thanks to @olikraus & @AnHardt for their help with this effort.

----

Short term to do list (about 2 days):
1. Finish adding SPI drivers
2. Implement I2C 
3. Move U8G2 driver setup out of ultralcd_impl_DOGM.h into a separate file.
4. CLEAN UP
5. Merge these changes with the current code base (will make it much easier to see what's going on

Mid range to do:
1. Work with Due & Teensy on compatibility.
2. Implement corrections/improvements as they come up

Long term to do (will need to find a resource that understands the following):
1. **dogm_font_data_6x9_marlin.h** didn't have a .FON file in the Marlin database so I grabbed one based on the description within the file.  @AnHardt has indicated that it may not contain everything that was in the Marlin 1.x version.  
2. Convert to the U8G2 UTF mapper.

----

Long term to do list

Bob-the-Kuhn committed:
Port Marlin 2.0.x to U8G2

delete scratch files

get rid of a couple more scrtatch files

fix travis errors

fix problem created by merge

added I2C support

Cleanup so changes are easier to see

change to manual rebase

cleanup after switching to new pin naming scheme - sort of works

Roxy-3D commented:
Bob...    A couple of observations from looking at your spreadsheet.

- Building u8g using the Atmega-2560 versions of the code with PlatformIO results in a small amount of code bloat.
- U8G2 is every so slightly slower on Atmega-2560 boards than u8G.  Probably not significant.
- I am surprised the Atmega-2560 versions of the code put less than a 10% burden on the processor.  I would have guessed it was higher than that.
- Initial thinking is:   If the U8G2 code is cleaner than the original library...   It might make sense to replace the existing library.

It is nice having real numbers to help the discussion!   Thank You Bob!

fiveangle commented:
Cool Bob !  Completely aside, I wonder if any cycles are saved when running Marlin with a graphics board using the “LCD Emulation Mode” of U8g2 vs straight native ? (I’d imagine not, although I suppose one could use an outboard MCU to offload a bit of the load from the Marlin MCU running “LCD” mode ?

But on the topic of whether to replace U8g with U8g2, have you considered how much work it might be to segment out the AVR MCUs to allow either ?

Only reason I ask is because it seems the most hamstrung 128k boards are the ones that have a graphics lcd (Anet being the most popular). They are already beyond the flash limits (requiring basic bed leveling, plus SD and a handful of nicety features disabled, in order to retain graphics use) so deciding to move to U8g2 is a decision to essentially put a fork in all the 128k boards.

Given the levity of it, should the Marlin dev community just announce that 1.1.x is the latest revision to support the 128k boards and that beginning with 2.x, users are on their own (although PRs will continue to be accepted by community members) ? Or perhaps just say “not officially supported” to set the correct tone, as to not get into a disagreement ? (Although what is even “officially supported” in open source community driven software anyway ? :)

It just seems there is more and more demand by people complaining that Marlin doesn’t fit on their 128k board. If support is withdrawn, they can still be helped along at a best effort capacity, but users won’t be expecting (or demanding) support for all the latest features of thier 128k board. 

This is just a feeler to see what others sentiment is. As an owner of one of these boards, I certainly appreciate that it continues to work as well as it does, but at some point people might prefer to have a realistic expectation set, rather that believe that they are being forsaken for no reason.

-=dave

Roxy-3D commented:
> It just seems there is more and more demand by people complaining that Marlin doesn’t fit on their 128k board. If support is withdrawn, they can still be helped along at a best effort capacity, but users won’t be expecting (or demanding) support for all the latest features of thier 128k board.

Yes...   But also remember there are very few people actually supporting Marlin's firmware.  It takes a lot of time to write, debug and test firmware.    Should that time be 'invested' in old, obsolete platforms that can be replaced for $20 ?

And the thing is this...   Nobody is saying they can't run Marlin on those 128 KB boards.   It might just be they can't run the latest and greatest version of Marlin on those boards with all the options they want.     Also, there is another possibility that hasn't been explored.   Are there big parts of UG82 that can be surgically cut out to shrink its program memory foot print?    

This is one of those topics where there isn't a 'right' answer.   Even with the 32-bit effort we have tried to set the 'right' tone by saying the initial 32-bit release should support the current 8-bit code as faithfully as possible.  And then going forward, we will try to add any new feature to both halves of the world.   We know at some point we will have to support both an 8-bit and 32-bit fork.   But we are trying to delay that point in time as much as reasonably possible.

psavva commented:
I think that Marlin running on 8 bit processors is very important.  A huge percentage of the 8bit setups rely on Marlin, and is probably the reason it's so heavily utilised by companies offering printers like the A8...

If you are to use U8g2, I'm guessing unbloating is certainly going to be needed to ensure backwards compatibility as Roxy suggested.

I do have another idea though...
Is there any sense in using both libraries? Maybe allowing limited 8 bit MCU's continue with u8g via the HAL, and only include U8g2 for beefier hardware also via HAL?

Roxy-3D commented:
> Is there any sense in using both libraries? Maybe allowing limited 8 bit MCU's continue with u8g via the HAL, and only include U8g2 for beefier hardware also via HAL?

It may be possible to do this.   We already do this to some degree on the Graphical LCD and the 20x4 Character  based LCD.   But if we can slim down the U8g2 library some what...   I would prefer to keep things simpler.   

Either way...  There is something else in the mix that might preempt everything.   Pretty much, everybody is in agreement the LCD interface for Marlin needs to be re-written.   Ideally, we would want to do that in a way where we don't care what LCD display is actually being used.   It maybe, this concern turns out to be a non-issue.    We don't actually use much of the u8g libraries.   It maybe with a new interface supporting call backs, we can just grab a few small sections of the code and support all processors and display combinations.

fiveangle commented:
> psavva: I do have another idea though... Is there any sense in using both libraries? 

What a great idea, why didn't I think of that 😉 

> fiveangle: But on the topic of whether to replace U8g with U8g2, have you considered how much work it might be to segment out the AVR MCUs to allow either ?

> psavva: I think that Marlin running on 8 bit processors is very important. 

I think you misunderstood.  The Ug8 vs. Ug82 debate isn't about dropping all 8-bit support: it's only about dropping 128k board support for the most useful configs.

This used to be **_Pick 2_**:

- Bed Levelling Support
- SD Support
- Graphics Panel Support.

Now it's already _**Pick 1-<sup>1</sup>/<sub>2</sub>**_ for any 128k board.

You have to give up other safety measures like BED thermal protection, prevent long extrusion, and a hand full of other little things like babystepping, etc, in order to get Bed Levelling and Panel Support on 128k.  It is so challenging, I had to compile every option and then delicately balance the value add of each when trying to fit it all in.  My Printrbot Simple Metal has BILINEAR ABL and a Ug8 panel and has litterally 347 bytes of PROGMEM available 🤣 and that's only after careful disabling of what I would call some critical safety features.  It's so challenging, I kept this older Configuration*.h around annotated with PROGMEM usage that I can point users to in order to help them understand the limits: https://github.com/fiveangle/Marlin/blob/progmem_usage_comments/Marlin/Configuration.h


How can things move forward without any place to store the new bits ?  Again, the question is not about supporting 8-bit boards, but **how can _128k boards_ be supported moving forward ?** Marlin confortably runs and fits on 8-bit 256k boards without issue (delta high-speed printing not withstanding :).

> roxy: if we can slim down the U8g2 library some what... I would prefer to keep things simpler.

Agree 100%

> roxy: everybody is in agreement the LCD interface for Marlin needs to be re-written... maybe, this concern turns out to be a non-issue... we can just grab a few small sections of the code and support all processors and display combinations.

Of course re-writing the entire graphics support works too - lol 👍 

-=dave

Roxy-3D commented:
> Of course re-writing the entire graphics support works too - lol :+1:

The good news is there are few people that enjoy doing the LCD work.   So the opportunity to start with a clean slate and do a re-write just might appeal to one of them.

AnHardt commented:
Thanks for the detail numbers and a better description about the circumstances measured with.
For a better understanding of what i'm talking about also see https://github.com/MarlinFirmware/Marlin/issues/8008#issuecomment-337433216
For the more technical interested / a brief look into the numbers.

Looking at Arduino Marlin 1.1.5 AVR u8g:
VIKI2s rendering time is not affected by screenrot.  Surprising - but an explanation for why Panucatt dared to mount it top down. Sorry for my misinterpretation of the numbers i likely got from RRDFG.
VIKI2 update time seems to be much improved compared to when i tested this.
The huge amount of sent bytes by the RRDFG compared to the others is because of sending a byte in two nibbles and transferring the target address multiple times (hardware requirement).

Comparing "Arduino Marlin 1.1.5 AVR u8g" to "PlatformIO AVR u8g"
Looking at the RRDFG EPROM and RAM usage seems to be decreased while speed seems to be reduced a bit. Is this from different GCC versions, different optimization settings or simply different code? Missing the "LCD not defined numbers". The differences in the VIKI2 sizes may be because of the altered device.
For the VIKI2 improved bit speed is not well corresponding to increased update times. Why is all that time lost in rendering? 
The difference in update time (71.7 vs 71.4) likely show the limitations of the system for time measurements. (So from now on mistrust all differences smaller then that)

Now for the most relevant part - Comparing  "PlatformIO AVR u8g" with  "PlatformIO AVR u8g²"
Again missing the data for "LCD not defined numbers" for u8g. We could see if parts of the code or libraries are compiled in when not needed.
We see a small increase of sizes. You already talked about.
We see a a bit faster VIKI2. Likely caused only by the much higher bit speed - rendering is slower. Moving to u8g² is a good deal for the VIKI2.
We see a much slower RRDFG. Only the numbers without rotation are relevant. (57.3 vs 79.0) is a major drawback That's about 25% slower than we are used to. So why is this?
Rendering time is about the same slower as the VIKI2 - maybe a bit less penalized.
Bitspeed is increased by a good amount (900K -> 1.2M ).  That may cause problems with some boards (pixel noise - shifted blocks).
Increased number of transferred bytes. That's likely a bug in u8g²! We know we can transfer the screens content with less than 2520bytes. 
A lower load% for the rotated screen is likely unreal when rendering and update lasts longer.

Looking at the numbers for the ARMs is pretty clueless. They are big and fast enough.
But we clearly see  RRDFG is at its hardware limits. Bit speed can be gradually tuned a bit, but will never exceed ~1.2M. (Hardware SPI with DMA from a big buffer may help)


Conclusion:
At the moment the penalty for the RRDFG is to high to dare the change to u8g². Fixing the too many transferred bytes will only gradually help. 

I concentrate on the RRDFG numbers because i strongly assume it's the by far most common graphical display out there. It's cheep, available since ages and you see it in about every Chinese printer with an AVR board. It's the display we have to care about.

@Blue-Marlin, right now, brought up the idea of using the ST7920 (display controller of the RRDFG) in character mode. 16 rows, 4 lines, 4 16x16 custom chars, maybe mixing of graphics and text at the same time, 128 ASCII symbols in 8x16pixel size, unbelievable amount of Chinese symbols in 16x16pixel - but then only 8x4 chars. That could reduce the amount of transferred bytes considerably - but for sure will be    as ugly as a lifejacket. (http://pdf1.alldatasheet.com/datasheet-pdf/view/326219/SITRONIX/ST7920.html)

Bob-the-Kuhn commented:
I've just added I2C support.  Mostly for completeness.

At 400KHz max bit rate they really suck up cpu cycles.

----

OK - 2560/1280 RRDFG code size is the most important followed by display update time.

I'll go back & see if there's anything I can do to shrink U8G2.  

I took a look at some older work where I added drivers to U8G (PR #7857)  It also had the lower code size.  If I can't make U8G2 smaller then I'll go back to the older work and bring it up to date.

AnHardt commented:
Hey @Bob-the-Kuhn.
From your code i guess you are using a logic analyser to get the timing. Does it have a SPI protocol analyser? Can the result be exported into a .txt or something else where i can search in? Would be nice if i could compare the output u8g vs. u8g² for the RRDFG to find what the additional transferred bytes are. 
And while you are at it - an additional test for the stock u8glib device for the RRDFG (U8GLIB_ST7920_128X64_4X u8g(LCD_PINS_D4, LCD_PINS_ENABLE, LCD_PINS_RS);) would be nice to compare ouer with the u8g device. I just need to know if the higher or lower amount of bytes is transferred or a protocol if it's a third number. :-)  

Analysing u8g² source for the complete sent sequence is, at the moment, too hard for me. While searching for some bytes more where you know what to search for sems to be much easier.

Don't put too much energy in it. If this isn't a easy task for you i'll hack the spi transfer functions to protocol what i need.

AnHardt commented:
In the special case of languages using only the first 127 chars (US-ASCII aka English) we could cut away all symbols above 127 in the font. Will save about 1K ROM. Will work for u8g and u8g².
In the same special case we could remove utf8 support. No clue about how much this will save and is much more work then the above. (I really don't want to cut my baby. ;-) )

Bob-the-Kuhn commented:
Yes, I have a cheap logic analyzer with a protocol analyzer.  The Saleae software + Hantek analyzer has been great.

Here's the RRDFG difference.
[RRDFG_dif_U8G vs U8G2.zip](https://github.com/MarlinFirmware/Marlin/files/1412180/RRDFG_dif_U8G.vs.U8G2.zip)

The differences in the 1024 byte (2048 SPI byte) data portions are:
- every 16 (32) display bytes there is one more "0xF8" in the page select sequence
- every 128 (256) display bytes there are three added bytes (0xF8, 0x30 & 0xE0) before the page select sequence.

thinkyhead commented:
> Now it's already Pick 1-1/2 for any 128k board.

I was able to enable the following options on a CR-10 (u8g v1) which uses a 1284p…
```
LEVEL_BED_CORNERS
EEPROM_SETTINGS
SDSUPPORT
SPEAKER
REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER
QUICK_HOME
STATUS_MESSAGE_SCROLLING
SDCARD_SORT_ALPHA (SDSORT_USES_RAM, SDSORT_CACHE_NAMES)
MENU_HOLLOW_FRAME
BABYSTEPPING
DOUBLECLICK_FOR_Z_BABYSTEPPING
ADVANCED_PAUSE_FEATURE
```
Disabled/enabled to get back some space & performance:
```
ARC_SUPPORT
NO_WORKSPACE_OFFSETS
```
The 1284p has lots of extra SRAM, so it's able to pre-allocate 256 slots for `SDCARD_SORT_ALPHA`.

thinkyhead commented on Marlin/src/lcd/ultralcd.cpp:
```diff
@@ -1082,17 +1082,16 @@ void kill_screen(const char* lcd_msg) {
-            u8g2.drawBitmap(left  + 20, 49 - dir, 2, 13, PSTR(down_arrow_bmp));
-          }
+        // Draw a representation of the nozzle
+        if (PAGE_CONTAINS(3, 16))  u8g2.drawBitmap(nozzle + 6, 4 - dir, 2, 12, PSTR(nozzle_bmp));
```
Is `PSTR()` needed here? I'm not sure what it does with pointers to things other than strings.

Roxy-3D replied:
Do we need a PPTR() ?   A Program Memory Based Pointer ???   My guess is PSTR handles an arbitrary pointer just fine.   But the name PSTR implies the generated pointer is to a text string, which isn't true.

thinkyhead replied:
If the u8g2 library expects these pointers to be `PROGMEM` then it already does the right thing, so `PSTR` is redundant. But if u8g2 expects normal SRAM pointers, maybe `PSTR` magically remaps them…? I'll wait and see whether `PSTR` was needed to make it work before assuming it's bad.

AnHardt replied:
It's the replacement for original constructs like
```
u8g.drawBitmapP(nozzle + 6, 4 - dir, 2, 12, nozzle_bmp);
```
Looks ok to me.

AnHardt replied:
Where nozzle_bmp is in PROGMEM.

thinkyhead replied:
So… `PSTR` is not needed then. GTK!

fiveangle commented:
> I was able to enable the following options on a CR-10 (u8g v1) which uses a 1284p…

We must all be doing something wrong then ! [grin] :)

thinkyhead commented:
> We must all be doing something wrong then

If you eat meat…

Otherwise, I think it was the extra disables that helped me fit all those in. After I did lots of testing, I enabled most of those features in the CR-10 example config that comes with Marlin. So far, no panic reports, so that's encouraging.

Bob-the-Kuhn commented:
@AnHardt - trimming **dogm_font_data_ISO10646_1.h** back to symbols 32-127 saves 1400 bytes.

There is not a UTF mapper for this font.  I commented out all the mappers and the image size stayed the same.

Next I'll see if we can save some more by re-writing the default Arduino U8G spi driver.

Bob-the-Kuhn commented:
> Next I'll see if we can save some more by re-writing the default Arduino U8G spi driver.

Didn't take too long to figure out that was a dead end.

I was able to trim the U8G2 code down but PlatformIO U8G is still has the smallest FLASH size with just the RRDFG display enabled.
```
  U8G PLATFORMIO    83808
  U8G2 PLATFORMIO   85150  +1342
  U8G Arduino       85844  +2036
```

----

Tomorrow I'll submit PRs to move the reduced font size into 1.1.x & 2.0.x

AnHardt commented:
@Bob-the-Kuhn 
Incredible comparison between u8g and u8g². Thanks!

My untested suggestions to avoid the extra bytes:
in `...\csrc\u8x8_d_st7920.c`
```diff
@@ -53,7 +53,7 @@ static const uint8_t u8x8_d_st7920_init_seq[] = {
   U8X8_C(0x002),                               /* disable scroll, enable CGRAM adress  */
   U8X8_C(0x001),                               /* clear RAM, needs 1.6 ms */
   U8X8_DLY(4),                                 /* delay 2ms */
-
+  //U8X8_C(0x03E),

   U8X8_END_TRANSFER(),                 /* disable chip */
   U8X8_END()                                   /* end of sequence */
@@ -63,6 +63,7 @@ static const uint8_t u8x8_d_st7920_powersave0_seq[] = {
   U8X8_START_TRANSFER(),               /* enable chip, delay is part of the transfer start */
   U8X8_C(0x038),                               /* 8 Bit interface (DL=1), basic instruction set (RE=0) */
   U8X8_C(0x00c),                               /* display on, cursor & blink off */
+  //U8X8_C(0x03E),
   U8X8_END_TRANSFER(),                 /* disable chip */
   U8X8_END()                                   /* end of sequence */
 };
@@ -71,6 +72,7 @@ static const uint8_t u8x8_d_st7920_powersave1_seq[] = {
   U8X8_START_TRANSFER(),               /* enable chip, delay is part of the transfer start */
   U8X8_C(0x038),                               /* 8 Bit interface (DL=1), basic instruction set (RE=0) */
   U8X8_C(0x008),                               /* display off */
+  //U8X8_C(0x03E),
   U8X8_END_TRANSFER(),                 /* disable chip */
   U8X8_END()                                   /* end of sequence */
 };
@@ -125,12 +127,13 @@ uint8_t u8x8_d_st7920_common(u8x8_t *u8x8, uint8_t msg, uint8_t arg_int, void *a
       ptr = ((u8x8_tile_t *)arg_ptr)->tile_ptr;        /* data ptr to the tiles */
       /* The following byte is sent to allow the ST7920 to sync up with the data */
       /* it solves some issues with garbage data */
-      u8x8_cad_SendCmd(u8x8, 0x03e );  /* enable extended mode */
+      u8x8_cad_SendCmd(u8x8, 0x03e );  /* enable extended mode */  // even this can likely be omitted when done in u8x8_d_st7920_init_seq[] and restored in u8x8_d_st7920_powersave1_seq[]
       for( i = 0; i < 8; i++ )
       {
-       u8x8_cad_SendCmd(u8x8, 0x03e ); /* enable extended mode */
+       //u8x8_cad_SendCmd(u8x8, 0x03e );       /* enable extended mode */ // this is the extra 3 byte sequence
        u8x8_cad_SendCmd(u8x8, 0x080 | (y+i) );      /* y pos  */
-       u8x8_cad_SendCmd(u8x8, 0x080 | x );      /* set x pos */
+       //u8x8_cad_SendCmd(u8x8, 0x080 | x );      /* set x pos */ // this is the extra 0xF8
+  U8X8_C(0x080 | x ),
        c = ((u8x8_tile_t *)arg_ptr)->cnt;      /* number of tiles */

        //u8x8->gpio_and_delay_cb(u8x8, U8X8_MSG_DELAY_NANO, 200, NULL);        /* extra dely required */
```
I really have to dig in the leftover boxes from moving to find my hardware for testing.

**EDIT:**  Ups. :-) My u8g² source was not that up to date. The 3byte sequence in the loop is already comment.
So it's the one directly above the loop. Makes even more sense - once a tile.
____________________________________________________

At least `ultralcd_st7920_u8glib_rrd.h` could be optimized for size, but that would be counterproductive - it's already optimized for speed.

Is that true? Compiled with 
* AVR RRDFG U8G Arduino

is now larger than
* AVR RRDFG U8G2 PLATFORMIO

In the .zip at the top this looked very different.
_____________________________________________________

Be careful to have two versions in `ISO10646_1.h` then. There are languages using the second part. 1400bytes are great.

Bob-the-Kuhn commented:
Do you know what languages use the second part?  Are they the ones that do not use the UTF mapper?

We could enable the reduced size font for just English.

AnHardt commented:
Where the font files are included

```cpp
#if ENABLED(DISPLAY_CHARSET_ISO10646_1) 
  #if ENABLED(MAPPER_C2C3)
    use the normal (long) version
  #else
    use the short version
  #endif
#endif

Bob-the-Kuhn commented:
I've verified which languages use the extended symbol set and which use the reduced symbols set.

PRs #8092 and #8095 have been submitted.

Bob-the-Kuhn commented:
I just tried to rebase this PR.  The result is not useable.

I'll have to see what this change to the pin/port numbering system is all about before I can fix things.

thinkyhead commented:
> I just tried to rebase this PR. The result is not useable.

Let me know if you're still stuck on that. I may be able to get you a working branch pretty quickly, as I do conflict resolutions all day long… and sometimes in my sleep.

thinkyhead commented:
> I just tried to rebase this PR. The result is not useable.

Here it is, rebased and usable…

https://github.com/thinkyhead/Marlin/tree/bf2_UTF82

```
git remote add thinkyhead git@github.com:thinkyhead/Marlin.git
git fetch thinkyhead
git checkout U8G2-port
git reset --hard thinkyhead/bf2_UTF82
git push -f origin
```

Bob-the-Kuhn commented:
I've gotten the SPI portion mostly working again.  There's a serious problem that sends it off into the woods.

I'm closing this PR because the U8G2 software costs more FLASH than the U8G software when compiling for the AVR boards.
